File: cache.go
package melange

import (
	"sync"
	"time"
)

// cacheKey uniquely identifies a permission check.
// All fields are required to form a unique key. Partial matches are not
// supported - the cache is exact-match only.
type cacheKey struct {
	SubjectType ObjectType
	SubjectID   string
	Relation    Relation
	ObjectType  ObjectType
	ObjectID    string
}

// cacheEntry stores the result of a permission check.
// Both successful and failed checks are cached, including errors.
// This prevents repeated queries for denied permissions.
type cacheEntry struct {
	allowed   bool
	err       error
	expiresAt time.Time // zero means no expiry
}

// Cache stores permission check results.
// It is safe for concurrent use from multiple goroutines.
//
// Implementations should cache both allowed and denied permissions, including
// errors. This reduces database load for repeated checks of denied access.
type Cache interface {
	// Get retrieves a cached permission check result.
	// Returns (allowed, err, found). If found is false, the entry doesn't exist or is expired.
	Get(subject Object, relation Relation, object Object) (allowed bool, err error, ok bool)

	// Set stores a permission check result in the cache.
	Set(subject Object, relation Relation, object Object, allowed bool, err error)
}

// CacheImpl is the default in-memory cache implementation with optional TTL.
// It uses a sync.RWMutex for goroutine safety. For high-contention scenarios,
// consider a sharded cache or external cache (Redis, etc.).
//
// The cache grows unbounded within its TTL window. For long-running processes
// with large permission sets, consider periodic clearing or TTL-based expiry.
type CacheImpl struct {
	mu    sync.RWMutex
	items map[cacheKey]cacheEntry
	ttl   time.Duration // 0 means no expiry
}

// CacheOption configures a Cache.
type CacheOption func(*CacheImpl)

// WithTTL sets the time-to-live for cache entries.
// Entries older than TTL are considered stale and will be re-checked.
// A TTL of 0 (default) means entries never expire within the cache's lifetime.
//
// Choose TTL based on permission volatility:
//   - Short TTL (seconds): Frequently changing permissions, high security
//   - Medium TTL (minutes): Typical web applications
//   - Long TTL or none: Near-static permissions, performance-critical paths
func WithTTL(ttl time.Duration) CacheOption {
	return func(c *CacheImpl) {
		c.ttl = ttl
	}
}

// NewCache creates a new permission cache.
// The cache is safe for concurrent use but scoped to a single process.
// For distributed systems, implement Cache with a distributed store.
func NewCache(opts ...CacheOption) *CacheImpl {
	c := &CacheImpl{
		items: make(map[cacheKey]cacheEntry),
	}
	for _, opt := range opts {
		opt(c)
	}
	return c
}

// Get retrieves a cached permission check result.
// Returns (allowed, err, found). If found is false, the entry doesn't exist or is expired.
func (c *CacheImpl) Get(subject Object, relation Relation, object Object) (bool, error, bool) {
	key := cacheKey{
		SubjectType: subject.Type,
		SubjectID:   subject.ID,
		Relation:    relation,
		ObjectType:  object.Type,
		ObjectID:    object.ID,
	}

	c.mu.RLock()
	entry, ok := c.items[key]
	c.mu.RUnlock()

	if !ok {
		return false, nil, false
	}

	// Check expiry if TTL is set
	if !entry.expiresAt.IsZero() && time.Now().After(entry.expiresAt) {
		// Entry expired, remove it
		c.mu.Lock()
		delete(c.items, key)
		c.mu.Unlock()
		return false, nil, false
	}

	return entry.allowed, entry.err, true
}

// Set stores a permission check result in the cache.
func (c *CacheImpl) Set(subject Object, relation Relation, object Object, allowed bool, err error) {
	key := cacheKey{
		SubjectType: subject.Type,
		SubjectID:   subject.ID,
		Relation:    relation,
		ObjectType:  object.Type,
		ObjectID:    object.ID,
	}

	entry := cacheEntry{
		allowed: allowed,
		err:     err,
	}

	if c.ttl > 0 {
		entry.expiresAt = time.Now().Add(c.ttl)
	}

	c.mu.Lock()
	c.items[key] = entry
	c.mu.Unlock()
}

// Size returns the number of entries in the cache.
// Useful for monitoring cache growth and memory usage.
func (c *CacheImpl) Size() int {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return len(c.items)
}

// Clear removes all entries from the cache.
// Useful for testing or when permission data changes globally
// (e.g., after schema migration or mass permission updates).
func (c *CacheImpl) Clear() {
	c.mu.Lock()
	c.items = make(map[cacheKey]cacheEntry)
	c.mu.Unlock()
}

// Ensure CacheImpl implements Cache.
var _ Cache = (*CacheImpl)(nil)

File: clients/typescript/src/cache.ts
/**
 * Caching for Melange authorization checks.
 *
 * This module provides cache interfaces and implementations for storing
 * permission check results to reduce database load.
 */

import type { Decision } from './types.js';

/**
 * Cache stores permission check results.
 *
 * Implementations should be safe for concurrent access if the Checker
 * is shared across requests. For request-scoped caching, create a new
 * Checker per request with a request-scoped cache.
 */
export interface Cache {
  /**
   * Get a cached decision.
   *
   * @param key - Cache key
   * @returns Cached decision, or undefined if not found or expired
   */
  get(key: string): Promise<Decision | undefined>;

  /**
   * Store a decision in the cache.
   *
   * @param key - Cache key
   * @param value - Decision to cache
   */
  set(key: string, value: Decision): Promise<void>;

  /**
   * Clear all cached entries.
   */
  clear(): Promise<void>;
}

/**
 * NoopCache is a no-op cache that never stores anything.
 *
 * This is the default cache implementation, suitable for applications
 * that don't want caching overhead or have other caching strategies.
 */
export class NoopCache implements Cache {
  async get(_key: string): Promise<Decision | undefined> {
    return undefined;
  }

  async set(_key: string, _value: Decision): Promise<void> {
    // no-op
  }

  async clear(): Promise<void> {
    // no-op
  }
}

/**
 * MemoryCache is a simple in-memory cache with TTL support.
 *
 * This cache stores decisions in a Map with time-based expiration.
 * It's suitable for single-instance applications or request-scoped caching.
 *
 * For multi-instance deployments, consider using a distributed cache
 * like Redis with a custom Cache implementation.
 *
 * @example
 * ```typescript
 * import { Checker, MemoryCache } from '@pthm/melange';
 * import { Pool } from 'pg';
 *
 * const pool = new Pool({ connectionString: process.env.DATABASE_URL });
 * const cache = new MemoryCache(60000); // 60 second TTL
 * const checker = new Checker(pool, { cache });
 *
 * // First check hits database
 * await checker.check(user, 'can_read', repo);
 *
 * // Second check within 60s uses cache
 * await checker.check(user, 'can_read', repo); // cached
 * ```
 */
export class MemoryCache implements Cache {
  private readonly cache = new Map<string, CacheEntry>();
  private readonly ttlMs: number;

  /**
   * Create a new MemoryCache.
   *
   * @param ttlMs - Time to live in milliseconds (default: 60000 = 1 minute)
   */
  constructor(ttlMs: number = 60000) {
    if (ttlMs <= 0) {
      throw new Error('TTL must be positive');
    }
    this.ttlMs = ttlMs;
  }

  async get(key: string): Promise<Decision | undefined> {
    const entry = this.cache.get(key);
    if (!entry) {
      return undefined;
    }

    // Check expiration
    if (Date.now() > entry.expiresAt) {
      this.cache.delete(key);
      return undefined;
    }

    return entry.value;
  }

  async set(key: string, value: Decision): Promise<void> {
    this.cache.set(key, {
      value,
      expiresAt: Date.now() + this.ttlMs,
    });
  }

  async clear(): Promise<void> {
    this.cache.clear();
  }

  /**
   * Get the number of entries in the cache.
   *
   * Note: This includes expired entries that haven't been accessed yet.
   */
  get size(): number {
    return this.cache.size;
  }
}

interface CacheEntry {
  value: Decision;
  expiresAt: number;
}
